examples/pmp_convor.py

```diff
@@ -0,0 +1,212 @@
+#       format_name: percent
+#       format_version: '1.3'
+#       jupytext_version: 1.13.7
+#   kernelspec:
+#     display_name: Python 3 (ipykernel)
+#     language: python
+#     name: python3
+# ---
+
+# %% [markdown]
+# We use PGMax to reimplement the binary deconvolution experiment presented in the Section 5.6 of the [Perturb-and-max-product (PMP)](https://proceedings.neurips.cc/paper/2021/hash/07b1c04a30f798b5506c1ec5acfb9031-Abstract.html) Neurips 2021 paper.
+#
+# The original implementation is available on the [GitHub repository of the paper.](https://github.com/vicariousinc/perturb_and_max_product/blob/master/experiments/exp6_convor.py)
+
+# %%
+import functools
+
+import jax
+import matplotlib.pyplot as plt
+import numpy as np
+from tqdm.notebook import tqdm
+
+from pgmax.factors import logical
+from pgmax.fg import graph, groups
+
+
+# %%
+def plot_images(images, display=True, nr=None):
+    "Useful function for visualizing several images"
+    n_images, H, W = images.shape
+    images = images - images.min()
+    images /= images.max() + 1e-10
+
+    if nr is None:
+        nr = nc = np.ceil(np.sqrt(n_images)).astype(int)
+    else:
+        nc = n_images // nr
+        assert n_images == nr * nc
+    big_image = np.ones(((H + 1) * nr + 1, (W + 1) * nc + 1, 3))
+    big_image[..., :3] = 0
+    big_image[:: H + 1] = [0.5, 0, 0.5]
+
+    im = 0
+    for r in range(nr):
+        for c in range(nc):
+            if im < n_images:
+                big_image[
+                    (H + 1) * r + 1 : (H + 1) * r + 1 + H,
+                    (W + 1) * c + 1 : (W + 1) * c + 1 + W,
+                    :,
+                ] = images[im, :, :, None]
+            im += 1
+
+    if display:
+        plt.figure(figsize=(10, 10))
+        plt.imshow(big_image, interpolation="none")
+    return big_image
+
+
+# %% [markdown]
+# ### Load the data
+
+# %% [markdown]
+# We extract the data from the PMP paper, only keeping the first 20 images here for the sake of speed.
+
+# %%
+data = np.load("example_data/conv_problem.npz")["X"]
+data = data[:20]
+
+_ = plot_images(data[:, 0], nr=4)
+
+# %% [markdown]
+# ### Construct variable grid, initialize factor graph, and add factors
+
+# %% [markdown]
+# We build a factor graph equivalent to the one in the PMP paper.
+
+# %%
+n_images, n_chan, im_height, im_width = data.shape
+n_feat, feat_height, feat_width = 5, 6, 6
+s_height = im_height - feat_height + 1
+s_width = im_width - feat_width + 1
+
+S = groups.NDVariableArray(num_states=2, shape=(n_images, n_feat, s_height, s_width))
+W = groups.NDVariableArray(
+    num_states=2, shape=(n_chan, n_feat, feat_height, feat_width)
+)
+SW = groups.NDVariableArray(
+    num_states=2,
+    shape=(n_images, n_chan, im_height, im_width, n_feat, feat_height, feat_width),
+)
+X = groups.NDVariableArray(num_states=2, shape=data.shape)
+
+fg = graph.FactorGraph(variables=dict(S=S, W=W, SW=SW, X=X))
+
+# %%
+variable_names_for_OR_factors = {}
+
+# Add ANDFactors
```

Comment:
Make this comment more descriptive (how the ANDFactors combine S and W into intermediate SW variables, what do these variables mean, etc.)

Reply:
Could you also make an inline comment here that although currently the factor adding is done with huge for loops we have plans in the future to make this more efficient with new factor groups?